File: ios/Sources/PamNative/Modules/DeviceModule.swift
import Foundation
import Network
import UIKit

final class DeviceModule: NativeModule, ClosableNativeModule {
    private let lock = NSLock()
    private var nextId = 1
    private var watches: [Int: DeviceWatch] = [:]

    func invoke(method: String, payload: Data, completion: @escaping ModuleCompletion) {
        switch method {
        case "status":
            snapshot(completion)
        case "watch":
            do { try start(payload, completion) }
            catch { completion(.failure, Data(error.localizedDescription.utf8)) }
        case "next":
            do { try watch(payload).channel.next(completion) }
            catch { completion(.failure, Data(error.localizedDescription.utf8)) }
        case "stop":
            do {
                stop(try id(payload))
                completion(.success, Data())
            } catch {
                completion(.failure, Data(error.localizedDescription.utf8))
            }
        default:
            completion(.failure, Data("Unknown device method \(method)".utf8))
        }
    }

    private func snapshot(_ completion: @escaping ModuleCompletion) {
        let monitor = NWPathMonitor()
        let queue = DispatchQueue(label: "dev.pam.native.device-status")
        monitor.pathUpdateHandler = { path in
            monitor.cancel()
            DispatchQueue.main.async {
                UIDevice.current.isBatteryMonitoringEnabled = true
                let type: Int64
                if path.status != .satisfied {
                    type = 1
                } else if path.usesInterfaceType(.wifi) {
                    type = 2
                } else if path.usesInterfaceType(.cellular) {
                    type = 3
                } else if path.usesInterfaceType(.wiredEthernet) {
                    type = 4
                } else {
                    type = 5
                }
                let charging = UIDevice.current.batteryState == .charging ||
                    UIDevice.current.batteryState == .full
                let result = (try? WireMap.encode([
                    "batteryLevel": .decimal(Double(UIDevice.current.batteryLevel)),
                    "charging": .flag(charging),
                    "networkType": .integer(type),
                    "expensiveNetwork": .flag(path.isExpensive),
                    "lowPowerMode": .flag(ProcessInfo.processInfo.isLowPowerModeEnabled),
                ])) ?? Data()
                completion(.success, result)
            }
        }
        monitor.start(queue: queue)
    }

    private func start(_ payload: Data, _ completion: @escaping ModuleCompletion) throws {
        let values = try WireMap.decode(payload)
        let interval = min(max(values["intervalMs"]?.deviceInteger ?? 1_000, 250), 60_000)
        let channel = WatchChannel()
        let timer = DispatchSource.makeTimerSource(queue: DispatchQueue.global(qos: .utility))
        lock.lock()
        let id = nextId
        nextId += 1
        watches[id] = DeviceWatch(channel: channel, timer: timer)
        lock.unlock()
        timer.schedule(deadline: .now(), repeating: .milliseconds(Int(interval)))
        timer.setEventHandler { [weak self, weak channel] in
            self?.snapshot { status, data in
                if status == .success { channel?.offer(data) }
            }
        }
        timer.resume()
        completion(.success, try WireMap.encode(["subscription": .integer(Int64(id))]))
    }

    private func watch(_ payload: Data) throws -> DeviceWatch {
        let id = try id(payload)
        lock.lock()
        defer { lock.unlock() }
        guard let watch = watches[id] else { throw DeviceModuleError("Unknown device subscription") }
        return watch
    }

    private func id(_ payload: Data) throws -> Int {
        let values = try WireMap.decode(payload)
        guard case let .integer(value)? = values["subscription"] else {
            throw DeviceModuleError("Missing device subscription")
        }
        return Int(value)
    }

    private func stop(_ id: Int) {
        lock.lock()
        let watch = watches.removeValue(forKey: id)
        lock.unlock()
        watch?.timer.cancel()
        watch?.channel.close()
    }

    func close() {
        lock.lock()
        let ids = Array(watches.keys)
        lock.unlock()
        ids.forEach { stop($0) }
    }
}

private struct DeviceWatch {
    let channel: WatchChannel
    let timer: DispatchSourceTimer
}

private struct DeviceModuleError: LocalizedError {
    let message: String
    init(_ message: String) { self.message = message }
    var errorDescription: String? { message }
}

private extension WireValue {
    var deviceInteger: Int64? {
        if case let .integer(value) = self { return value }
        return nil
    }
}

File: android/app/src/main/java/dev/pam/nativeapp/modules/DeviceModule.kt
package dev.pam.nativeapp.modules

import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.os.BatteryManager
import android.os.PowerManager
import android.os.Handler
import android.os.Looper
import dev.pam.nativeapp.protocol.WireMap
import dev.pam.nativeapp.protocol.WireValue

import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger

internal class DeviceModule(private val context: Context) : NativeModule, AutoCloseable {
    private val main = Handler(Looper.getMainLooper())
    private val nextId = AtomicInteger(1)
    private val watches = ConcurrentHashMap<Int, DeviceWatch>()

    override fun invoke(method: String, payload: ByteArray, completion: ModuleCompletion) {
        runCatching {
            when (method) {
                "status" -> completion.complete(ModuleResultStatus.SUCCESS, statusPayload())
                "watch" -> start(payload, completion)
                "next" -> watch(payload).channel.next(completion)
                "stop" -> {
                    stop(id(payload))
                    completion.complete(ModuleResultStatus.SUCCESS, ByteArray(0))
                }
                else -> error("Unknown device method $method")
            }
        }.onFailure {
            completion.complete(ModuleResultStatus.FAILURE, (it.message ?: "Device status failed").toByteArray())
        }
    }

    private fun start(payload: ByteArray, completion: ModuleCompletion) {
        val interval = ((WireMap.decode(payload)["intervalMs"] as? WireValue.Integer)?.value ?: 1_000)
            .coerceIn(250, 60_000)
        val id = nextId.getAndIncrement()
        val channel = WatchChannel()
        lateinit var runnable: Runnable
        runnable = Runnable {
            if (!watches.containsKey(id)) return@Runnable
            channel.offer(statusPayload())
            main.postDelayed(runnable, interval)
        }
        watches[id] = DeviceWatch(channel, runnable)
        main.post(runnable)
        completion.complete(
            ModuleResultStatus.SUCCESS,
            WireMap.encode(mapOf("subscription" to WireValue.Integer(id.toLong()))),
        )
    }

    private fun statusPayload(): ByteArray {
        val battery = context.registerReceiver(null, IntentFilter(Intent.ACTION_BATTERY_CHANGED))
        val level = battery?.getIntExtra(BatteryManager.EXTRA_LEVEL, -1) ?: -1
        val scale = battery?.getIntExtra(BatteryManager.EXTRA_SCALE, -1) ?: -1
        val status = battery?.getIntExtra(BatteryManager.EXTRA_STATUS, -1) ?: -1
        val charging = status == BatteryManager.BATTERY_STATUS_CHARGING ||
            status == BatteryManager.BATTERY_STATUS_FULL
        val connectivity = context.getSystemService(ConnectivityManager::class.java)
        val capabilities = connectivity.getNetworkCapabilities(connectivity.activeNetwork)
        val type = when {
            capabilities == null -> 1
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) -> 2
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) -> 3
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET) -> 4
            else -> 5
        }
        val expensive = capabilities?.hasCapability(
            NetworkCapabilities.NET_CAPABILITY_NOT_METERED,
        ) == false
        val lowPower = context.getSystemService(PowerManager::class.java).isPowerSaveMode
        return WireMap.encode(mapOf(
            "batteryLevel" to WireValue.Decimal(
                if (level >= 0 && scale > 0) level.toDouble() / scale else -1.0,
            ),
            "charging" to WireValue.Flag(charging),
            "networkType" to WireValue.Integer(type.toLong()),
            "expensiveNetwork" to WireValue.Flag(expensive),
            "lowPowerMode" to WireValue.Flag(lowPower),
        ))
    }

    private fun watch(payload: ByteArray): DeviceWatch =
        watches[id(payload)] ?: error("Unknown device subscription")

    private fun id(payload: ByteArray): Int =
        ((WireMap.decode(payload)["subscription"] as? WireValue.Integer)?.value
            ?: error("Missing device subscription")).toInt()

    private fun stop(id: Int) {
        watches.remove(id)?.let {
            main.removeCallbacks(it.runnable)
            it.channel.close()
        }
    }

    override fun close() {
        watches.keys.toList().forEach(::stop)
    }

    private data class DeviceWatch(
        val channel: WatchChannel,
        val runnable: Runnable,
    )
}
